core/src/main/java/jenkins/slaves/DefaultJnlpSlaveReceiver.java

```diff
@@ -124,7 +124,8 @@ public void afterProperties(@NonNull JnlpConnectionState event) {
         Channel ch = computer.getChannel();
         if (ch != null) {
             String cookie = event.getProperty(JnlpConnectionState.COOKIE_KEY);
-            if (cookie != null && MessageDigest.isEqual(cookie.getBytes(StandardCharsets.UTF_8), ch.getProperty(COOKIE_NAME).toString().getBytes(StandardCharsets.UTF_8))) {
+            String channelCookie = (String)ch.getProperty(JnlpConnectionState.COOKIE_KEY);
+            if (cookie != null && channelCookie != null && MessageDigest.isEqual(cookie.getBytes(StandardCharsets.UTF_8), channelCookie.getBytes(StandardCharsets.UTF_8))) {
```

Comment:
One of the confusing bits is that this would previously throw a NullPointerException.

Reply:
If it evet reached this check, yes...

Reply:
It did reach this check, throwing a NullPointerException, before these fixes. That was confusing but wasn't really the problem. With these changes, it doesn't trip over this in normal, valid situations, but it may still reduce confusion in some invalid reconnection situations.